libs/storage/src/single_object_store.cpp

```diff
@@ -0,0 +1,211 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2020 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "storage/single_object_store.hpp"
+#include "storage/storage_exception.hpp"
+
+static constexpr char const *LOGGING_NAME = "SingleObjectStore";
+
+namespace fetch {
+
+namespace platform {
+enum
+{
+  LITTLE_ENDIAN_MAGIC = 2337
+};
+}
+
+namespace storage {
+
+struct FileMetadata
+{
+  uint16_t magic;
+  uint16_t version;
+  uint64_t object_size;
+
+  bool Valid() const
+  {
+    return magic == platform::LITTLE_ENDIAN_MAGIC;
+  }
+
+  void Initialise(uint16_t version_to_set)
+  {
+    magic   = platform::LITTLE_ENDIAN_MAGIC;
+    version = version_to_set;
+  }
+};
+
+bool SingleObjectStore::Load(std::string const &file_name)
+{
+  file_name_ = file_name;
+
+  file_handle_.open(file_name, std::fstream::in | std::fstream::out | std::fstream::binary);
+
+  // If does not exist
+  if (!file_handle_)
+  {
+    file_handle_.open(file_name, std::fstream::in | std::fstream::out | std::fstream::binary |
+                                     std::fstream::trunc);
+  }
+
+  if (!(file_handle_ && file_handle_.is_open()))
+  {
+    FETCH_LOG_WARN(LOGGING_NAME, "Failed to open: ", file_name);
+    return false;
+  }
+
+  // To remain POD, this must be initialized rather than constructed
+  FileMetadata meta{};
+  meta.Initialise(version_);
+  static_assert(std::is_pod<FileMetadata>::value, "FileMetadata must be POD");
+
+  // Check it is either empty or non-corrupted
+  if (FileSize() == 0)
+  {
+    // Write metadata to new file
+    file_handle_.write(reinterpret_cast<char const *>(&meta), sizeof(meta));
+    file_handle_.flush();
+    return true;
+  }
+
+  if (FileSize() < sizeof(meta))
+  {
+    FETCH_LOG_WARN(
+        LOGGING_NAME,
+        "Attempted to open a file that had nonzero size but less than expected metadata");
+    return false;
+  }
+
+  // Read the metadata
+  file_handle_.seekg(0, std::fstream::beg);
+  file_handle_.read(reinterpret_cast<char *>(&meta), sizeof(meta));
```

Comment:
As @wkaluza mentioned in my PR - it makes sense to add check to the fail and bad bits here